[Site Name] Keep vertical input focused on back and reset the Site Intent screen on skip

ovitrif opened the pull request:
Alternative solution to: #16378

Internal ref: `p5T066-3a9-p2#comment-12083`

Proposes a different approach for handling the input focus on the `Site Intent screen` and hiding/showing the keyboard when navigating back from the `Site Name screen`.

<details open>
<summary>Preview</summary>

<img width="314" src="https://user-images.githubusercontent.com/4588074/164539776-5832030c-136a-4ecb-a4e5-292827de6b7e.gif">
</details>

<details>
<summary>1. Toggle ON the Site Name feature</summary>

1. From My Site Go to Me → App Settings → Debug settings
2. Scroll to Features in development section and tap on SiteNameFeatureConfig
3. Tap RESTART THE APP button
4. Since this feature is A/B tested go to **Abacus**, find the `wpandroid_site_name_v1` experiment and assign to your test a8c-user the treatment variation

**Alternatively hardcode the feature on by returning `true` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/521e6b3a296339bec026f8ee2648eab326d31129/WordPress/src/main/java/org/wordpress/android/util/config/SiteNameFeatureConfig.kt#L19)** 
</details>

2. Start the Site Creation flow
3. Choose an intent for your site from the default or proceed with a custom one
4. Tap the `←` back button
5. `Verify` that **the keyboard is shown and the input is focused**.
   Typing on the keyboard should work as expected = change the value of the input and trigger the search.
7. Tap `Skip`
4. Tap the `←` back button
5. `Verify` that **the keyboard is hidden and the screen UI is reset** to its initial state
   (equal to how it looks when starting the Site Creation Flow)
7. Keep doing exploratory testing by repeating steps `3 - 5` to check that it works as expected in all cases.

## Regression Notes
1. Potential unintended areas of impact
   N/a

2. What I did to test those areas of impact (or what existing automated tests I relied on)
   N/a

3. What automated tests I added (or what prevented me from doing so)
  - I added unit tests in `SiteCreationIntentsViewModelTest` to cover the updated logic.
  - 


PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

antonis committed:
Hides the keyboard when navigating back from the site name screen

ovitrif committed:
Handle keyboard visibility in fragment of intents screen

ovitrif committed:
Add unit tests for the new ui state update

ovitrif committed:
Suppress TooManyFunctions check in SiteCreationIntentsViewModel

ovitrif committed:
Reformat code in SiteCreationIntentsViewModel

ovitrif committed:
Fix introduced bug with typing custom vertical

ovitrif committed:
Merge pull request #16378 from wordpress-mobile/fix/site-name-hide-keyboard-on-back

[Site Name] Hides keyboard on back navigation

ovitrif commented:
@kyleaparker I added you to the reviewers for a design review and feedback on the new behavior.

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/SiteCreationIntentsFragment.kt:
```diff
@@ -78,11 +82,28 @@ class SiteCreationIntentsFragment : Fragment() {
+        uiState.vertical
+                ?.let {
+                    input.requestFocus()
+                    ActivityUtils.showKeyboard(input)
+                }
+                ?: run {
```
I have a slight concern about the behavior if `showKeyboard` ever returns `null`, since I guess this will also invoke the chained run block :thinking:

ovitrif replied:
That's a valid remark @mkevins! Thank you for sharing it 🙇‍♂️.

I agree we don't want changes to the return value of `ActivityUtils.showKeyboard` to influence this behavior 👍.

To make the code safer I wrapped the calls into an `also` block which makes sure to return the value of `uiState.vertical` in https://github.com/wordpress-mobile/WordPress-Android/pull/16380/commits/ae0528c57df33ae3f6bd024df232c8b900dc7253.

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/SiteCreationIntentsFragment.kt:
```diff
@@ -78,11 +82,28 @@ class SiteCreationIntentsFragment : Fragment() {
     }
 
+    private fun TextInputEditText.setTextUnobserved(text: CharSequence) {
+        removeTextChangedListener(inputWatcher)
+        setText(text)
+        addTextChangedListener(inputWatcher)
```
I _think_ this is safe, since I guess `setTextUnobserved` would only ever run _after_ the `onViewStateRestored`, but maybe we can add a `null` check just in case, to guard unexpected races where `inputWatcher` is `null` here? Wdyt?

ovitrif replied:
I wrapped the code that works with the inputWatcher in a `inputWatcher?.let {}` block for this in [3642566](https://github.com/wordpress-mobile/WordPress-Android/pull/16380/commits/364256618ca387ad01578987a24f31cad202bfd1), thank you for this remark also, I agree safer is better 🎉 .

mkevins reviewed (commented):
I tested this and it is working as described. During exploratory testing, there is one thing I found _slightly_ odd, not a problem, but just something I noticed: when I select an intent from the choices provided and then tap back, that intent is displayed in the input box (and focused) as though I typed it - but the list is not filtered as it would be had I actually typed it. This isn't really a problem, but just an observation I had that felt ever-so-slightly inconsistent.

Otoh, filtering would be strange as well, since the user would tap back presumably to _change_ their selection, and having it filtered to only the one they just picked would add extra steps for them to get back to the full list (i.e. deleting the text). :man_shrugging: 

I wonder if we should retain the search query (or custom intent) as is, rather than resetting to the chosen intent? :thinking:

ovitrif committed:
Avoid unexpected behavior if showKeyboard ever returns null

ovitrif committed:
Handle case when inputWatcher may be null silently

ovitrif commented:
> Otoh, filtering would be strange as well, since the user would tap back presumably to change their selection, and having it filtered to only the one they just picked would add extra steps for them to get back to the full list (i.e. deleting the text). 🤷‍♂️

I gave this a thought myself and kinda concluded in the end for the sake of having a working example to go forward and propose it as is to get more feedback 🙇

Imho, same as you mentioned, the user-friendly assumption we make here is that the user would want to change the vertical, thus presenting the default list is not bad although a bit inconsistent 😊

> I wonder if we should retain the search query (or custom intent) as is, rather than resetting to the chosen intent? 🤔

This might be a good improvement to the new behavior introduced in this PR! Thank you @mkevins 🙇 .

I actually think it will feel more polished by applying this. Otherwise we get a bit of a strange behavior with this example scenario:
- Search for `renting`
- select `Parenting`
- go back
- tap backspace on the keyboard
- `check` that the custom vertical becomes `Parentin` instead of `Renting` as it was initially. → This could be improved if we retain the search query instead.

If I understand correctly it will only apply when we have a search query or custom intent, and otherwise (when we're just selecting a vertical from the default list) it could stay as it is?

ovitrif committed:
Persist search query instead of selected vertical when it exists

ovitrif committed:
Fix failing test for skip tap

ovitrif committed:
Add test for persisting search query when it exists

ovitrif committed:
Rename vertical to input value

ovitrif committed:
Rename inputValue to retainedInputValue in ui state

ovitrif commented:
> I wonder if we should retain the search query (or custom intent) as is, rather than resetting to the chosen intent? 🤔

I updated the logic to retain the search query (or custom intent) instead of resetting the input value to the chosen intent and personally I like this behavior more.

The commits are: [2409163](https://github.com/wordpress-mobile/WordPress-Android/pull/16380/commits/240916342ceebbbb25a23e0d7e409132104eb080), and a fix because it was buggy when the searchQuery was empty or blank: [41d2185](https://github.com/wordpress-mobile/WordPress-Android/pull/16380/commits/41d21854ecf63f6738c0e1a6d79ecca44da78cff).

I also renamed the name of the property to `retainedInputValue` in hope of avoiding confusion when the code is read (having `inputValue` and `searchQuery` in the same data class was a bit confusing imho, and `vertical` was wrong since we're now persisting the searchQuery when it exists).

antonis reviewed (commented):
Thank you for exploring this alternative @ovitrif 🙇 
The code looks good but I'm a bit skeptical on which approach is better in terms of functionality and user experience.

The downside/bug of the current approach is that it would require from the user to refocus the search input when getting back to the search screen from the site name. The main downside of the alternative is that the users get back to a different screen from the one they started in some cases and this might be confusing. Another thing is that most users (according to the analytics up to now) choose one of the default intents and never see the search screen.

I tried to enumerate the different cases below. 

<details>
<summary>Current approach</summary>

https://user-images.githubusercontent.com/304044/164680862-f103fc1f-73f3-4535-ae7c-43e95281a290.mp4
</details>

1. Tap on a default intent -> back -> default screen ✅ 
2. Focus search -> enter text -> select -> back -> search screen
   * Keyboard closed ❌ 
3. Focus search -> enter text -> skip -> back -> search screen
   * Keyboard closed ❌ 
4. Focus search -> skip -> back -> search screen
   * Keyboard closed ❌ 
5. Skip -> back -> default screen ✅ 

<details>
<summary>Alternative</summary>

https://user-images.githubusercontent.com/304044/164680896-2576c5db-5ef9-4738-8d7d-3f720da43191.mp4
</details>

1. Tap on a default intent -> back -> search screen
   * Getting back to a screen not seen before ❌ 
   * The list is not filtered with the input ❌ 
2. Focus search -> enter text -> select -> back -> search screen ✅ 
3. Focus search -> enter text -> skip -> back -> default screen
   * Getting back to the initial screen which might be confusing ❌ 
   * Input is lost ❌ 
4. Focus search -> skip -> back -> search screen
   * Getting back to the initial screen which might be confusing ❌ 
5. Skip -> back -> default screen ✅ 

My suggestion would be to keep the current simpler approach for `19.7`. We can then get feedback from @kyleaparker since this involves UX and target the next release with a better fix. Wdyt?

ovitrif commented:
Thank you @antonis for your review & feedback!

> My suggestion would be to keep the current simpler approach for `19.7`. We can then get feedback from @kyleaparker since this involves UX and target the next release with a better fix. Wdyt?

I agree with this and I'm pinging @kyleaparker (did this initially but then remembered by default she's AFK on Fridays) so I thought if we'd be to target `19.7` we won't have enough time to wait for more feedback.

I'm also going to add the **[Needs Design Review]** status label and target `19.8` for now.

Let's collect more feedback since all concerns you mentioned are valid and it looks like we'd have to polish this a bit more if we're going to merge it 🙇

kyleaparker commented:
In the case of a custom input, this approach makes sense to me where we return the user to the filled in input. I could see people making spelling errors with custom inputs and wanting to correct those. With the defaults, I think it feels more natural to return to the initial state, it felt odd to see "Lifestyle" in the input after initially tapping on lifestyle from the main screen. Could it work to take a different approach depending if someone entered a custom vertical vs selecting a default?

ovitrif commented:
> In the case of a custom input, this approach makes sense to me where we return the user to the filled in input. I could see people making spelling errors with custom inputs and wanting to correct those. With the defaults, I think it feels more natural to return to the initial state, it felt odd to see "Lifestyle" in the input after initially tapping on lifestyle from the main screen. Could it work to take a different approach depending if someone entered a custom vertical vs selecting a default?

Thank you @kyleaparker for your feedback here 🙇 !

It does feel like a good adjustment to make it only for the user-typed custom input value, I'll check out how it goes like that and come back with a screen recording & code update!

ovitrif commented:
I'm closing this since it's stale for almost 2 months and I don't see when we'll progress on it.
We can always come back and review how to properly address the two issues.
